drivers/virtio/net/virtio_net.c

```diff
@@ -41,24 +42,26 @@
 #include <virtio/virtqueue.h>
 #include <virtio/virtio_net.h>
 
-/**
- * VIRTIO_PKT_BUFFER_LEN = VIRTIO_NET_HDR + ETH_HDR + ETH_PKT_PAYLOAD_LEN
- * VIRTIO_NET_HDR: 10 bytes in length in legacy mode + 2 byte of padded data.
- *		   12 bytes in length in modern mode.
- */
-#define VIRTIO_HDR_LEN          12
-#define VIRTIO_PKT_BUFFER_LEN ((UK_ETH_PAYLOAD_MAXLEN) \
-			       + (UK_ETH_HDR_UNTAGGED_LEN) \
-			       + (VIRTIO_HDR_LEN))
+#define DRIVER_NAME	"virtio-net"
 
-#define DRIVER_NAME           "virtio-net"
+/* VIRTIO_PKT_BUFFER_LEN = VIRTIO_NET_HDR + ETH_HDR + ETH_PKT_PAYLOAD_LEN */
+#define VIRTIO_PKT_BUFFER_LEN(_vndev)				\
+	((UK_ETH_PAYLOAD_MAXLEN) + (UK_ETH_HDR_UNTAGGED_LEN) +	\
+	 (size_t)virtio_net_hdr_size(_vndev))
 
+/**
+ * When mergeable buffers are not negotiated, the virtio_net_hdr_padded struct
+ * below is placed at the beginning of the netbuf data. Use 4 bytes of pad to
+ * both keep the VirtIO header and the data non-contiguous and to keep the
+ * frame's payload 4 byte aligned.
+ */
+#define VTNET_HDR_SIZE_PADDED				\
+	(ALIGN_UP(sizeof(struct virtio_net_hdr), 4) + 4)
```

Comment:
I like this, but could we have `ALIGN_UP( (size_t)virtio_net_hdr_size(_vndev), 4)` instead? With our current supported virtio features, we only need 10-12 bytes for the virtio header. Meanwhile `struct virtio_net_hdr` has 20 bytes. Feel free to discard this comment if it's too much of a hassle for 8 bytes.

Reply:
I thought about it at some point too, but in the end didn't do it. But since you also brought it up, I'll change it 👍🏼